src/algebra/group_power/order.lean

```diff
@@ -13,82 +13,111 @@ Note that some lemmas are in `algebra/group_power/lemmas.lean` as they import fi
+  rw [pow_add, pow_succ', mul_assoc, ← pow_succ],
+  exact lt_mul_of_one_lt_right' _ (one_lt_pow' ha k.succ_ne_zero)
+end
 
-end add_monoid
+end preorder
 
-section add_group
-variable [ordered_add_comm_group A]
+section linear_order
 
-theorem gsmul_nonneg {a : A} (H : 0 ≤ a) {n : ℤ} (hn : 0 ≤ n) :
-  0 ≤ n • a :=
-begin
-  lift n to ℕ using hn,
-  rw gsmul_coe_nat,
-  apply nsmul_nonneg H,
-end
+variables [monoid M] [linear_order M] [covariant_class M M (*) (≤)]
 
-end add_group
+@[to_additive nsmul_nonneg_iff]
+lemma one_le_pow_iff {x : M} {n : ℕ} (hn : n ≠ 0) : 1 ≤ x ^ n ↔ 1 ≤ x :=
+⟨le_imp_le_of_lt_imp_lt $ λ h, pow_lt_one' h hn, λ h, one_le_pow_of_one_le' h n⟩
 
-section cancel_add_monoid
-variable [ordered_cancel_add_comm_monoid A]
+@[to_additive nsmul_nonpos_iff]
+lemma pow_le_one_iff {x : M} {n : ℕ} (hn : n ≠ 0) : x ^ n ≤ 1 ↔ x ≤ 1 :=
+@one_le_pow_iff (order_dual M) _ _ _ _ _ hn
 
-theorem nsmul_lt_nsmul {a : A} {n m : ℕ} (ha : 0 < a) (h : n < m) :
-  n • a < m • a :=
-let ⟨k, hk⟩ := nat.le.dest h in
+@[to_additive nsmul_pos_iff]
+lemma one_lt_pow_iff {x : M} {n : ℕ} (hn : n ≠ 0) : 1 < x ^ n ↔ 1 < x :=
+lt_iff_lt_of_le_iff_le (pow_le_one_iff hn)
+
+@[to_additive nsmul_neg_iff]
+lemma pow_lt_one_iff {x : M} {n : ℕ} (hn : n ≠ 0) : x ^ n < 1 ↔ x < 1 :=
+lt_iff_lt_of_le_iff_le (one_le_pow_iff hn)
+
+@[to_additive nsmul_eq_zero_iff]
+lemma pow_eq_one_iff {x : M} {n : ℕ} (hn : n ≠ 0) : x ^ n = 1 ↔ x = 1 :=
+by simp only [le_antisymm_iff, pow_le_one_iff hn, one_le_pow_iff hn]
+
+end linear_order
+
+section group
+
+variables [group G] [preorder G] [covariant_class G G (*) (≤)]
+
+@[to_additive gsmul_nonneg]
+theorem one_le_gpow {x : G} (H : 1 ≤ x) {n : ℤ} (hn : 0 ≤ n) :
+  1 ≤ x ^ n :=
 begin
-  have succ_swap : n.succ + k = n + k.succ := nat.succ_add n k,
-  calc n • a = (n • a : A) + (0 : A) : (add_zero _).symm
-    ... < n • a + (k.succ • a : A) : add_lt_add_left (nsmul_pos ha (nat.succ_pos k)) _
-    ... = m • a : by rw [← hk, succ_swap, add_nsmul]
+  lift n to ℕ using hn,
+  rw gpow_coe_nat,
+  apply one_le_pow_of_one_le' H,
 end
 
-end cancel_add_monoid
+end group
 
 namespace canonically_ordered_comm_semiring
-variable [canonically_ordered_comm_semiring R]
 
-theorem pow_pos {a : R} (H : 0 < a) : ∀ n : ℕ, 0 < a ^ n
-| 0     := by { nontriviality, rw pow_zero, exact zero_lt_one }
-| (n+1) := by { rw pow_succ, exact mul_pos.2 ⟨H, pow_pos n⟩ }
+variables [canonically_ordered_comm_semiring R]
 
-@[mono] lemma pow_le_pow_of_le_left {a b : R} (hab : a ≤ b) : ∀ i : ℕ, a^i ≤ b^i
-| 0     := by simp
-| (k+1) := by { rw [pow_succ, pow_succ],
-    exact mul_le_mul' hab (pow_le_pow_of_le_left k) }
-
-theorem one_le_pow_of_one_le {a : R} (H : 1 ≤ a) (n : ℕ) : 1 ≤ a ^ n :=
-by simpa only [one_pow] using pow_le_pow_of_le_left H n
-
-theorem pow_le_one {a : R} (H : a ≤ 1) (n : ℕ) : a ^ n ≤ 1:=
-by simpa only [one_pow] using pow_le_pow_of_le_left H n
+theorem pow_pos {a : R} (H : 0 < a) (n : ℕ) : 0 < a ^ n :=
+pos_iff_ne_zero.2 $ pow_ne_zero _ H.ne'
```

Comment:
Does this generalize to `monoid_with_zero` and some co(ntra)variant classes?

Reply:
Do we have a class of this kind that implies `0 ≤ 1`?

Reply:
Unfortunately not. We have it for `linear_ordered_comm_monoid_with_zero`, but here we don't have to assume linear ordered-ness.